### DISCUSSION TITLE
mem32[] as default argument: should it work?

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#11050] [RP2040 / Pico] [answered]
I was debugging a problem with a function that read the global enable register of the PWMs in the Pico. I found the `mem32[]` call in the function default argument to be the culprit.

```
from machine import PWM, Pin, mem32
from micropython import const

REG_BASE = const(0x40050000)
GLOBAL_EN = const(REG_BASE + 0xA0)

def tst(en = mem32[GLOBAL_EN] & 0xFF):
    print("tst() input argument: " + hex(en))
    print("tst() read inside: " + hex(mem32[GLOBAL_EN] & 0xFF))

tst(0xA5)

pwm1 = PWM(Pin(0))
pwm1.freq(1000)
pwm1.duty_u16(1000)

tst()

pwm2 = PWM(Pin(14))
pwm2.freq(1000)
pwm2.duty_u16(1000)

tst()

```

Here the result of running it (either Thonny or importing on repl). `mem32[]` returns 0 when called as the default.

```
tst() input argument: 0xa5
tst() read inside: 0x0
tst() input argument: 0x0
tst() read inside: 0x1
tst() input argument: 0x0
tst() read inside: 0x81
```

Am I doing something wrong?

**Comment by robert-hh** [comment, marked as answer]
Argument defaults are evaluated only once at compilation/parsing time. That's written somewhere, and may be even Python standard.

**Reply by cve2022** [reply]
Hello, Robert, in hindsight it makes much sense.

Thank you.
